internal/receiver/smartagentreceiver/filtering.go

```diff
@@ -0,0 +1,255 @@
+func (mf *monitorFiltering) HasEnabledMetricInGroup(group string) bool {
+	if mf.metadata == nil {
+		return false
+	}
+
+	for _, m := range mf.EnabledMetrics() {
+		if mf.metadata.Metrics[m].Group == group {
+			return true
+		}
+	}
+	return false
+}
+
+// HasAnyExtraMetrics returns true if there is any custom metric
+// enabled for this output instance.
+func (mf *monitorFiltering) HasAnyExtraMetrics() bool {
+	return mf.hasExtraMetrics
+}
+
+func buildFilterSet(metadata *monitors.Metadata, conf config.MonitorCustomConfig) (*dpfilters.FilterSet, error) {
+	coreConfig := conf.MonitorConfigCore()
+
+	filter, err := coreConfig.FilterSet()
+	if err != nil {
+		return nil, err
+	}
+
+	excludeFilters := []dpfilters.DatapointFilter{filter}
+
+	// If sendAll is true or there are no metrics don't bother setting up
+	// filtering
+	if metadata != nil && len(metadata.Metrics) > 0 && !metadata.SendAll {
+		// Make a copy of extra metrics from config so we don't alter what the user configured.
+		extraMetrics := append([]string{}, coreConfig.ExtraMetrics...)
+
+		// Monitors can add additional extra metrics to allow through such as based on config flags.
+		if monitorExtra, ok := conf.(config.ExtraMetrics); ok {
+			extraMetrics = append(extraMetrics, monitorExtra.GetExtraMetrics()...)
+		}
+
+		includedMetricsFilter, err := newMetricsFilter(metadata, extraMetrics, coreConfig.ExtraGroups)
+		if err != nil {
+			return nil, fmt.Errorf("unable to construct extraMetrics filter: %s", err)
+		}
+
+		// Prepend the included metrics filter.
+		excludeFilters = append([]dpfilters.DatapointFilter{dpfilters.Negate(includedMetricsFilter)}, excludeFilters...)
+	}
+
+	filterSet := &dpfilters.FilterSet{
+		ExcludeFilters: excludeFilters,
+	}
+
+	return filterSet, nil
+}
+
+var _ dpfilters.DatapointFilter = &extraMetricsFilter{}
+
+// Filter of datapoints based on included status and user configuration of
+// extraMetrics and extraGroups.
+type extraMetricsFilter struct {
+	metadata     *monitors.Metadata
+	extraMetrics map[string]bool
+	stringFilter *filter.BasicStringFilter
+}
+
+func validateMetricName(metadata *monitors.Metadata, metricName string) error {
+	if strings.TrimSpace(metricName) == "" {
+		return errors.New("metric name cannot be empty")
+	}
+
+	if metadata.SendUnknown {
+		// The metrics list isn't exhaustive so can't do extra validation.
+		return nil
+	}
+
+	if strings.ContainsRune(metricName, '*') {
+		// Make sure a metric pattern matches at least one metric.
+		m, err := filter.NewBasicStringFilter([]string{metricName})
+		if err != nil {
+			return err
+		}
+
+		for metric := range metadata.Metrics {
+			if m.Matches(metric) {
+				return nil
+			}
+		}
+
+		logrus.Warnf("extraMetrics: metric pattern '%s' did not match any available metrics for monitor %s",
+			metricName, metadata.MonitorType)
+	}
+
+	if !metadata.HasMetric(metricName) {
+		logrus.Warnf("extraMetrics: metric '%s' does not exist for monitor %s", metricName, metadata.MonitorType)
```

Comment:
We should be using the receiver's zap logger and not the global logrus logger.

Reply:
addressed